### PR TITLE
test: test fsevents error reporting

### PR DESCRIPTION
**Pull request by indutny**
Create a lot of `FSEventStream`s to make `FSEventStreamStart()` fail.
Fetch the error code and verify that it is correct.

/cc @bnoordhuis

### CODE REVIEW

File: test/task.h [hunk 1]
```diff
@@ -153,6 +158,24 @@ enum test_status {
     return TEST_SKIP;                                                         \
   } while (0)
 
+#if !defined(_WIN32)
+
+# define TEST_FILE_LIMIT(num)                                                 \
```
Comment (bnoordhuis) [inline test/task.h:163]: I don't think this should be in the header, it's only used in one place.

File: test/test-fs-event.c [hunk 2]
```diff
@@ -556,3 +556,110 @@ static void fs_event_cb_close(uv_fs_event_t* handle, const char* filename,
...
+  ASSERT(handle != NULL);
+  close_cb_called++;
+  free(handle);
+}
+
+
+TEST_IMPL(fs_event_error_reporting) {
+  unsigned int i;
+  uv_loop_t* loops[1024];
+  uv_fs_event_t* events[ARRAY_SIZE(loops)];
```
Comment (bnoordhuis) [inline test/test-fs-event.c]: You can just allocate these on the stack and not bother with malloc/free, right?

File: test/test-fs-event.c [hunk 3]
```diff
@@ -556,3 +556,110 @@ static void fs_event_cb_close(uv_fs_event_t* handle, const char* filename,
...
+    ASSERT(0 == uv_timer_init(loop, &timer));
+    ASSERT(0 == uv_timer_start(&timer, timer_cb_nop, 2, 0));
+    uv_run(loop, UV_RUN_DEFAULT);
+    ASSERT(1 == timer_cb_called && 1 == close_cb_called);
+    if (fs_event_error_reported != 0)
+      break;
+  }
+
+  /* At least one loop should fail */
+  ASSERT(fs_event_error_reported == -EMFILE);
```
Comment (bnoordhuis) [inline test/test-fs-event.c]: UV_EMFILE?

File: test/test-fs-event.c [hunk 4]
```diff
@@ -556,3 +556,110 @@ static void fs_event_cb_close(uv_fs_event_t* handle, const char* filename,
...
+
+  /* Stop and close all events, and destroy loops */
+  while (1) {
```
Comment (bnoordhuis) [inline test/test-fs-event.c]: do/while loop?

File: test/test-fs-event.c [hunk 5]
```diff
@@ -556,3 +556,110 @@ static void fs_event_cb_close(uv_fs_event_t* handle, const char* filename,
...
+
+    ASSERT(0 == uv_fs_event_stop(event));
+    uv_ref((uv_handle_t*) event);
+    uv_close((uv_handle_t*) event, fs_event_error_report_close_cb);
+
+    close_cb_called = 0;
+    uv_run(loop, UV_RUN_DEFAULT);
+    ASSERT(close_cb_called == 1);
+
+    uv_loop_delete(loops[i]);
```
Comment (bnoordhuis) [inline test/test-fs-event.c]: loop

File: test/test-fs-event.c [hunk 6]
```diff
@@ -556,3 +556,110 @@ static void fs_event_cb_close(uv_fs_event_t* handle, const char* filename,
...
+                                  fs_event_error_report_cb,
+                                  "watch_dir",
+                                  0));
+    uv_unref((uv_handle_t*) event);
+
+    /* Let loop run for some time */
+    ASSERT(0 == uv_timer_init(loop, &timer));
+    ASSERT(0 == uv_timer_start(&timer, timer_cb_nop, 2, 0));
+    uv_run(loop, UV_RUN_DEFAULT);
+    ASSERT(1 == timer_cb_called && 1 == close_cb_called);
```
Comment (bnoordhuis) [inline test/test-fs-event.c]: Split these into two asserts.

File: test/test-fs-event.c (hunk 2 above)
Comment (indutny) [inline test/test-fs-event.c]: Won't this be quite a lot for a stack?

Comment (bnoordhuis) [inline test/test-fs-event.c]: How big is one element?  Even if it's 1 kB, that's still only 1 MB of stack space.

File: test/task.h (hunk 1 above)
Comment (indutny) [inline test/task.h:163]: In two places, actually. I was planning to replace it in another test later on.

File: test/test-fs-event.c (hunk 2 above)
Comment (indutny) [inline test/test-fs-event.c]: Ok, if you wish so.

**Comment by indutny** [comment]
@bnoordhuis pushed fixes.

File: test/test-fs-event.c [hunk 7]
```diff
@@ -556,3 +556,110 @@ static void fs_event_cb_close(uv_fs_event_t* handle, const char* filename,
...
+    ASSERT(0 == uv_timer_start(&timer, timer_cb_nop, 2, 0));
+    uv_run(loop, UV_RUN_DEFAULT);
+    ASSERT(1 == timer_cb_called);
+    ASSERT(1 == close_cb_called);
+    if (fs_event_error_reported != 0)
+      break;
+  }
+
+  /* At least one loop should fail */
+  ASSERT(fs_event_error_reported == -UV_EMFILE);
```
Comment (bnoordhuis) [inline test/test-fs-event.c]: This can't possibly work...

File: test/test-fs-event.c (hunk 5 above)
Comment (bnoordhuis) [inline test/test-fs-event.c]: Not fixed yet.

File: test/test-fs-event.c [hunk 8]
```diff
@@ -556,3 +556,110 @@ static void fs_event_cb_close(uv_fs_event_t* handle, const char* filename,
...
+    ASSERT(close_cb_called == 1);
+
+    uv_loop_delete(loops[i]);
+
+    loops[i] = NULL;
+    if (i == 0)
+      break;
+    else
+      i--;
+  } while(1);
```
Comment (bnoordhuis) [inline test/test-fs-event.c]: This is not quite what I mean.  E.g. `} while (i-- != 0);` was what I had in mind.

**Comment by indutny** [comment]
@bnoordhuis pushed fixes.

File: test/test-fs-event.c [hunk 9]
```diff
@@ -556,3 +556,106 @@ static void fs_event_cb_close(uv_fs_event_t* handle, const char* filename,
...
+    uv_close((uv_handle_t*) event, fs_event_error_report_close_cb);
+
+    close_cb_called = 0;
+    uv_run(loop, UV_RUN_DEFAULT);
+    ASSERT(close_cb_called == 1);
+
+    uv_loop_delete(loop);
+
+    loops[i] = NULL;
+  } while(i-- != 0);
```
Comment (bnoordhuis) [inline test/test-fs-event.c:645]: Style: missing space

File: test/task.h [hunk 10]
```diff
@@ -153,6 +158,24 @@ enum test_status {
...
+    do {                                                                      \
+      struct rlimit lim;                                                      \
+      lim.rlim_cur = (num);                                                   \
+      lim.rlim_max = (num);                                                   \
```
Comment (bnoordhuis) [inline test/task.h:167]: You're evaluating num twice here.

**Comment by indutny** [comment]
Landed with fixes in 86b5c1e. Thank you!

**Pull request closed without merging**